Error while checking build/O2/fullCI for 5f4b3754269503b4e0143ac2fa10ba77b6efe2aa at 2024-11-02 21:49:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
command /sw/slc8_x86-64/O2/13649-slc8_x86-64-local1/prodtests/full-system-test/dpl-workflow.sh had nonzero exit code 1
[ERROR] error parsing options of o2-eve-export-workflow: unrecognised option '--number-of_tracks'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/01a1d8ea5911db33a93bce4383a12ab5a84aa553/slc8_x86-64/o2checkcode/1.0-local389/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13649/5f4b3754269503b4e0143ac2fa10ba77b6efe2aa/build_O2_fullCI/pretty.html).

--- a/Detectors/GlobalTrackingWorkflow/src/ReaderDriverSpec.cxx
+++ b/Detectors/GlobalTrackingWorkflow/src/ReaderDriverSpec.cxx
@@ -75,7 +75,7 @@ void ReadeDriverSpec::run(ProcessingContext& pc)
     LOGP(debug, "OUTVEC {}/{}", v.back().getMin().asString(), v.back().getMax().asString());
   }
   count++;
-  if ((HBFINI::LastIRFrameIndex == -1 && count == mNTF) || (v.size() && v.back().isLast())) {
+  if ((HBFINI::LastIRFrameIndex == -1 && count >= mNTF) || (v.size() && v.back().isLast())) {
     pc.services().get<ControlService>().endOfStream();
     pc.services().get<ControlService>().readyToQuit(QuitRequest::Me);
   }

--- a/Detectors/Raw/src/HBFUtilsInitializer.cxx
+++ b/Detectors/Raw/src/HBFUtilsInitializer.cxx
@@ -58,6 +58,7 @@ HBFUtilsInitializer::HBFUtilsInitializer(const o2f::ConfigContext& configcontext
       bool helpasked = configcontext.helpOnCommandLine(); // if help is asked, don't take for granted that the ini file is there, don't produce an error if it is not!
       auto conf = configcontext.options().isSet(HBFConfOpt) ? configcontext.options().get<std::string>(HBFConfOpt) : "";
       if (!conf.empty()) {
+        int nopts = 0;
         auto vopts = o2::utils::Str::tokenize(conf, ',');
         for (const auto& optStr : vopts) {
           if (optStr == UpstreamOpt) {
@@ -65,6 +66,7 @@ HBFUtilsInitializer::HBFUtilsInitializer(const o2f::ConfigContext& configcontext
             continue;
           }
           HBFOpt opt = getOptType(optStr);
+          nopts++;
           if ((opt == HBFOpt::INI || opt == HBFOpt::JSON) && !helpasked) {
             o2::conf::ConfigurableParam::updateFromFile(optStr, "HBFUtils", true); // update only those values which were not touched yet (provenance == kCODE)
             const auto& hbfu = o2::raw::HBFUtils::Instance();
@@ -76,7 +78,12 @@ HBFUtilsInitializer::HBFUtilsInitializer(const o2f::ConfigContext& configcontext
             hbfuInput = optStr;
           } else if (opt == HBFOpt::ROOT) {
             rootFileInput = optStr;
+          } else {
+            LOGP(fatal, "uknown hbfutils-config option {}", optStr);
           }
+        }
+        if (!nopts && !helpasked) {
+          LOGP(fatal, "No source was provided for timing input of --hbfutils-config");
         }
       }
       done = true;

--- a/EventVisualisation/Base/CMakeLists.txt
+++ b/EventVisualisation/Base/CMakeLists.txt
@@ -17,9 +17,12 @@ o2_add_library(EventVisualisationBase
                        src/GeometryManager.cxx
                        src/FileWatcher.cxx
                        src/DirectoryLoader.cxx
-
+                       src/EveConfParam.cxx
         PUBLIC_LINK_LIBRARIES ROOT::Eve
                                      O2::CCDB
                                      O2::EventVisualisationDataConverter
                                      O2::DetectorsBase
 )
+
+o2_target_root_dictionary(EventVisualisationBase
+                          HEADERS include/EventVisualisationBase/EveConfParam.h)

--- a/EventVisualisation/Base/include/EventVisualisationBase/EveConfParam.h
+++ b/EventVisualisation/Base/include/EventVisualisationBase/EveConfParam.h
@@ -1,0 +1,48 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef ALICE_O2_EVENTVISUALISATION_CONFIG_H
+#define ALICE_O2_EVENTVISUALISATION_CONFIG_H
+#include "CommonUtils/ConfigurableParam.h"
+#include "CommonUtils/ConfigurableParamHelper.h"
+
+namespace o2::event_visualisation
+{
+struct EveConfParam : o2::conf::ConfigurableParamHelper<EveConfParam> {
+  float EMCCellTimeMax = 100.f;  // Max. EMCAL cell time (in ns)
+  float EMCCellEnergyMin = 0.3f; // Min. EMCAL cell energy (in GeV)
+
+  float timeMinMax[2] = {-999.f, -999.f}; // if min<max (microseconds), show only track in this range of every TF
+
+  int maxFiles = 150;     // maximum number of json files in folder
+  int maxTracks = -1;     // maximum number of track stored in json file (-1 means no limit)
+  int maxBytes = 3000000; // number of bytes stored in time interval which stops producing new data file (-1 means no limit)
+
+  int maxPVs = -1;                              // if > 0: enable PV mode and limit max number of vertices
+  int minTracks = -1;                           // don't create file if less than the specified number of all tracks is present
+  int minITSTracks = -1;                        // don't create file if less than the specified number of ITS tracks is present
+  int onlyNthEvent = -1;                        // process only every nth event (if > 1)
+  float PVXYZMin[3] = {-999.f, -999.f, -999.f}; // primary vertex min x,y,z
+  float PVXYZMax[3] = {999.f, 999.f, 999.f};    // primary vertex max x,y,z
+  float TPCOnlyMaxDCARZ[2] = {-999.f, -999.f};  // max DCA r,z (if positive) for TPC only tracks in the primary vertex mode
+  float TPCEtaAbsMax = -1.f;                    // if positive, show TPC tracks only below this abs eta
+
+  bool PVMode = false;         // produce jsons with individual primary vertices, not total time frame data
+  bool PVTriggersMode = false; // instead of drawing vertices with tracks (and maybe calorimeter triggers), draw vertices with calorimeter triggers (and maybe tracks)"
+  bool trackSorting = false;   // sort track by track time before applying filters
+  bool filterITSROF = false;   // don't display tracks outside ITS readout frame
+  bool calibrateEMC = true;    // apply on-the-fly EMCAL calibration
+
+  O2ParamDef(EveConfParam, "eveconf");
+};
+} // namespace o2::event_visualisation
+
+#endif

--- a/EventVisualisation/Base/src/EveConfParam.cxx
+++ b/EventVisualisation/Base/src/EveConfParam.cxx
@@ -1,0 +1,14 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "EventVisualisationBase/EveConfParam.h"
+
+O2ParamImpl(o2::event_visualisation::EveConfParam);

--- a/EventVisualisation/Base/src/EventVisualisationBaseLinkDef.h
+++ b/EventVisualisation/Base/src/EventVisualisationBaseLinkDef.h
@@ -1,0 +1,22 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifdef __CLING__
+
+#pragma link off all globals;
+#pragma link off all classes;
+#pragma link off all functions;
+
+#pragma link C++ class o2::event_visualisation::EveConfParam + ;
+
+#pragma link C++ class o2::conf::ConfigurableParamHelper < o2::event_visualisation::EveConfParam> + ;
+
+#endif

--- a/EventVisualisation/Workflow/include/EveWorkflow/EveWorkflowHelper.h
+++ b/EventVisualisation/Workflow/include/EveWorkflow/EveWorkflowHelper.h
@@ -29,6 +29,7 @@
 #include "ITSBase/GeometryTGeo.h"
 #include "TPCFastTransform.h"
 #include "TPCReconstruction/TPCFastTransformHelperO2.h"
+#include "DataFormatsTPC/VDriftCorrFact.h"
 #include "Framework/AnalysisDataModel.h"
 #include "DetectorsVertexing/PVertexerParams.h"
 
@@ -124,27 +125,17 @@ class EveWorkflowHelper
 
   using AODMFTTracks = aod::MFTTracks;
   using AODMFTTrack = AODMFTTracks::iterator;
-
-  enum Filter : uint8_t {
-    ITSROF,
-    TimeBracket,
-    EtaBracket,
-    TotalNTracks,
-    NFilters
-  };
-
-  using FilterSet = std::bitset<Filter::NFilters>;
-
   using Bracket = o2::math_utils::Bracketf_t;
 
-  EveWorkflowHelper(const FilterSet& enabledFilters = {}, std::size_t maxNTracks = -1, const Bracket& timeBracket = {}, const Bracket& etaBracket = {}, bool primaryVertexMode = false);
+  EveWorkflowHelper();
   static std::vector<PNT> getTrackPoints(const o2::track::TrackPar& trc, float minR, float maxR, float maxStep, float minZ = -25000, float maxZ = 25000);
+  void setTPCVDrift(const o2::tpc::VDriftCorrFact* v);
   void selectTracks(const CalibObjectsConst* calib, GID::mask_t maskCl, GID::mask_t maskTrk, GID::mask_t maskMatch);
   void selectTowers();
   void setITSROFs();
   void addTrackToEvent(const o2::track::TrackPar& tr, GID gid, float trackTime, float dz, GID::Source source = GID::NSources, float maxStep = 4.f);
   void draw(std::size_t primaryVertexIdx, bool sortTracks);
-  void drawTPC(GID gid, float trackTime);
+  void drawTPC(GID gid, float trackTime, float dz = 0.f);
   void drawITS(GID gid, float trackTime);
   void drawMFT(GID gid, float trackTime);
   void drawMCH(GID gid, float trackTime);
@@ -170,7 +161,7 @@ class EveWorkflowHelper
   void drawMFTTrack(GID gid, o2::track::TrackParFwd track, float trackTime);
   void drawForwardTrack(GID gid, mch::TrackParam track, float startZ, float endZ, float trackTime);
   void drawITSClusters(GID gid);
-  void drawTPCClusters(GID gid);
+  void drawTPCClusters(GID gid, float trackTimeTB = -2.e9); // if trackTimeTB<-1.e9, then use tpcTrack.getTime0()
   void drawMFTClusters(GID gid);
   void drawMCHClusters(GID gid);
   void drawMIDClusters(GID gid);
@@ -194,17 +185,14 @@ class EveWorkflowHelper
 
   void save(const std::string& jsonPath, const std::string& ext, int numberOfFiles);
 
-  FilterSet mEnabledFilters;
-  std::size_t mMaxNTracks;
-  Bracket mTimeBracket;
-  Bracket mEtaBracket;
-  bool mPrimaryVertexMode;
+  bool mUseTimeBracket = false;
+  bool mUseEtaBracketTPC = false;
+  Bracket mTimeBracket{};
+  Bracket mEtaBracketTPC;
   const o2::globaltracking::RecoContainer* mRecoCont = nullptr;
   const o2::globaltracking::RecoContainer* getRecoContainer() const { return mRecoCont; }
   void setRecoContainer(const o2::globaltracking::RecoContainer* rc) { mRecoCont = rc; }
   void setEMCALCellRecalibrator(o2::emcal::CellRecalibrator* calibrator) { mEMCALCalib = calibrator; }
-  void setMaxEMCALCellTime(float maxtime) { mEMCALMaxCellTime = maxtime; }
-  void setMinEMCALCellEnergy(float minenergy) { mEMCALMinCellEnergy = minenergy; }
   TracksSet mTrackSet;
   o2::event_visualisation::VisualisationEvent mEvent;
   std::unordered_map<GID, std::size_t> mTotalDataTypes;
@@ -220,13 +208,12 @@ class EveWorkflowHelper
   o2::phos::Geometry* mPHOSGeom;
   o2::emcal::Geometry* mEMCALGeom;
   o2::emcal::CellRecalibrator* mEMCALCalib = nullptr;
-
-  float mMUS2TPCTimeBins = 5.0098627;
+  const o2::tpc::VDriftCorrFact* mTPCVDrift = nullptr;
+  float mMUS2TPCTimeBins = 5.0098627f;
+  float mTPCTimeBins2MUS = 0.199606f;
   float mITSROFrameLengthMUS = 0; ///< ITS RO frame in mus
   float mMFTROFrameLengthMUS = 0; ///< MFT RO frame in mus
   float mTPCBin2MUS = 0;
-  float mEMCALMaxCellTime = 100.;  ///< EMCAL cell time cut (in ns)
-  float mEMCALMinCellEnergy = 0.3; ///< EMCAL cell energy cut (in GeV)
   static int BCDiffErrCount;
   const o2::vertexing::PVertexerParams* mPVParams = nullptr;
 };

--- a/EventVisualisation/Workflow/include/EveWorkflow/O2DPLDisplay.h
+++ b/EventVisualisation/Workflow/include/EveWorkflow/O2DPLDisplay.h
@@ -18,6 +18,7 @@
 
 #include "ReconstructionDataFormats/GlobalTrackID.h"
 #include "DataFormatsGlobalTracking/RecoContainer.h"
+#include "TPCCalibration/VDriftHelper.h"
 #include "DetectorsBase/GRPGeomHelper.h"
 #include "EMCALCalib/CellRecalibrator.h"
 #include "EMCALWorkflow/CalibLoader.h"
@@ -58,14 +59,9 @@ class O2DPLDisplaySpec : public o2::framework::Task
                    std::shared_ptr<o2::base::GRPGeomRequest> gr,
                    std::shared_ptr<o2::emcal::CalibLoader> emcCalibLoader,
                    const std::string& jsonPath, const std::string& ext,
-                   std::chrono::milliseconds timeInterval, int numberOfFiles, int numberOfTracks, int numberOfBytes,
-                   bool eveHostNameMatch, int minITSTracks, int minTracks, bool filterITSROF, bool filterTime,
-                   const EveWorkflowHelper::Bracket& timeBracket, bool removeTPCEta,
-                   const EveWorkflowHelper::Bracket& etaBracket, bool trackSorting, int onlyNthEvent,
-                   bool primaryVertex, int maxPrimaryVertices, bool primaryVertexTriggers,
-                   float primaryVertexMinZ, float primaryVertexMaxZ, float primaryVertexMinX, float primaryVertexMaxX, float primaryVertexMinY, float primaryVertexMaxY,
-                   float maxEMCALCellTime, float minEMCALCellEnergy)
-    : mDisableWrite(disableWrite), mUseMC(useMC), mTrkMask(trkMask), mClMask(clMask), mDataRequest(dataRequest), mGGCCDBRequest(gr), mEMCALCalibLoader(emcCalibLoader), mJsonPath(jsonPath), mExt(ext), mTimeInterval(timeInterval), mNumberOfFiles(numberOfFiles), mNumberOfTracks(numberOfTracks), mNumberOfBytes(numberOfBytes), mEveHostNameMatch(eveHostNameMatch), mMinITSTracks(minITSTracks), mMinTracks(minTracks), mFilterITSROF(filterITSROF), mFilterTime(filterTime), mTimeBracket(timeBracket), mRemoveTPCEta(removeTPCEta), mEtaBracket(etaBracket), mTrackSorting(trackSorting), mOnlyNthEvent(onlyNthEvent), mPrimaryVertexMode(primaryVertex), mMaxPrimaryVertices(maxPrimaryVertices), mPrimaryVertexTriggers(primaryVertexTriggers), mPrimaryVertexMinZ(primaryVertexMinZ), mPrimaryVertexMaxZ(primaryVertexMaxZ), mPrimaryVertexMinX(primaryVertexMinX), mPrimaryVertexMaxX(primaryVertexMaxX), mPrimaryVertexMinY(primaryVertexMinY), mPrimaryVertexMaxY(primaryVertexMaxY), mEMCALMaxCellTime(maxEMCALCellTime), mEMCALMinCellEnergy(minEMCALCellEnergy), mRunType(o2::parameters::GRPECS::NONE)
+                   std::chrono::milliseconds timeInterval,
+                   bool eveHostNameMatch)
+    : mDisableWrite(disableWrite), mUseMC(useMC), mTrkMask(trkMask), mClMask(clMask), mDataRequest(dataRequest), mGGCCDBRequest(gr), mEMCALCalibLoader(emcCalibLoader), mJsonPath(jsonPath), mExt(ext), mTimeInterval(timeInterval), mEveHostNameMatch(eveHostNameMatch), mRunType(o2::parameters::GRPECS::NONE)
 
   {
     this->mTimeStamp = std::chrono::high_resolution_clock::now() - timeInterval; // first run meets condition
@@ -82,32 +78,10 @@ class O2DPLDisplaySpec : public o2::framework::Task
   bool mDisableWrite = false; // skip writing result (for testing performance)
   bool mUseMC = false;
   bool mEveHostNameMatch;                  // empty or correct hostname
-  int mMinITSTracks;                       // minimum number of ITS tracks to produce a file
-  int mMinTracks;                          // minimum number of all tracks to produce a file
-  bool mFilterITSROF;                      // don't display tracks outside ITS readout frame
-  bool mFilterTime;                        // don't display tracks outside [min, max] range in TF time
-  bool mRemoveTPCEta;                      // don't display TPC tracks inside [min, max] eta range
-  bool mPrimaryVertexMode;                 // produce files per primary vertex
-  EveWorkflowHelper::Bracket mTimeBracket; // [min, max] range in TF time for the filter
-  EveWorkflowHelper::Bracket mEtaBracket;  // [min, max] eta range for the TPC tracks removal
   std::string mJsonPath;                   // folder where files are stored
   std::string mExt;                        // extension of created files (".json" or ".root")
   std::chrono::milliseconds mTimeInterval; // minimal interval between files in milliseconds
-  int mNumberOfFiles;                      // maximum number of files in folder - newer replaces older
-  int mNumberOfTracks;                     // maximum number of track in single file (0 means no limit)
-  int mNumberOfBytes;                      // number of bytes stored in period which causes stopping saving a new file
-  bool mTrackSorting;                      // perform sorting tracks by track time before applying filters
-  int mOnlyNthEvent;                       // process only every nth event.
-  int mMaxPrimaryVertices;                 // max number of primary vertices to draw per time frame
   bool mPrimaryVertexTriggers;             // instead of drawing vertices with tracks (and maybe calorimeter triggers), draw vertices with calorimeter triggers (and maybe tracks)
-  float mPrimaryVertexMinZ;                // minimum z position of the primary vertex
-  float mPrimaryVertexMaxZ;                // maximum z position of the primary vertex
-  float mPrimaryVertexMinX;                // minimum x position of the primary vertex
-  float mPrimaryVertexMaxX;                // maximum x position of the primary vertex
-  float mPrimaryVertexMinY;                // minimum y position of the primary vertex
-  float mPrimaryVertexMaxY;                // maximum y position of the primary vertex
-  float mEMCALMaxCellTime;                 // max abs EMCAL cell time (in ns)
-  float mEMCALMinCellEnergy;               // min EMCAL cell energy (in GeV)
   int mEventCounter = 0;
   std::chrono::time_point<std::chrono::high_resolution_clock> mTimeStamp;
 
@@ -119,6 +93,7 @@ class O2DPLDisplaySpec : public o2::framework::Task
   std::shared_ptr<o2::base::GRPGeomRequest> mGGCCDBRequest;
   std::shared_ptr<o2::emcal::CalibLoader> mEMCALCalibLoader;
   std::unique_ptr<o2::emcal::CellRecalibrator> mEMCALCalibrator;
+  o2::tpc::VDriftHelper mTPCVDriftHelper{};
 };
 
 } // namespace o2::event_visualisation

--- a/EventVisualisation/Workflow/src/EveWorkflowHelper.cxx
+++ b/EventVisualisation/Workflow/src/EveWorkflowHelper.cxx
@@ -14,6 +14,7 @@
 
 #include <EveWorkflow/EveWorkflowHelper.h>
 #include "EventVisualisationBase/ConfigurationManager.h"
+#include "EventVisualisationBase/EveConfParam.h"
 #include "EventVisualisationDataConverter/VisualisationEventSerializer.h"
 #include "ReconstructionDataFormats/GlobalTrackID.h"
 #include "ReconstructionDataFormats/VtxTrackRef.h"
@@ -22,6 +23,7 @@
 #include "ITStracking/IOUtils.h"
 #include "MFTTracking/IOUtils.h"
 #include "DataFormatsGlobalTracking/RecoContainerCreateTracksVariadic.h"
+#include "ReconstructionDataFormats/PrimaryVertex.h"
 #include "DetectorsBase/GRPGeomHelper.h"
 #include "DetectorsBase/Propagator.h"
 #include "TPCBase/ParameterElectronics.h"
@@ -152,6 +154,8 @@ bool EveWorkflowHelper::isInsideTimeBracket(float t)
 void EveWorkflowHelper::selectTracks(const CalibObjectsConst* calib,
                                      GID::mask_t maskCl, GID::mask_t maskTrk, GID::mask_t maskMatch)
 {
+  const auto& conf = EveConfParam::Instance();
+
   auto correctTrackTime = [this](auto& _tr, float t0, float terr) {
     if constexpr (isTPCTrack<decltype(_tr)>()) {
       // unconstrained TPC track, with t0 = TrackTPC.getTime0+0.5*(DeltaFwd-DeltaBwd) and terr = 0.5*(DeltaFwd+DeltaBwd) in TimeBins
@@ -193,8 +197,7 @@ void EveWorkflowHelper::selectTracks(const CalibObjectsConst* calib,
       }
     }
   };
-
-  auto creator = [maskTrk, this, &correctTrackTime, &flagTime, &fixMFTMCHMIDLabel](auto& trk, GID gid, float time, float terr) {
+  auto creator = [&conf, maskTrk, this, &correctTrackTime, &flagTime, &fixMFTMCHMIDLabel](auto& trk, GID gid, float time, float terr) {
     fixMFTMCHMIDLabel(gid);
 
     const auto src = gid.getSource();
@@ -206,11 +209,11 @@ void EveWorkflowHelper::selectTracks(const CalibObjectsConst* calib,
 
     auto bracket = correctTrackTime(trk, time, terr);
 
-    if (mEnabledFilters.test(Filter::TimeBracket) && !isInsideTimeBracket(bracket)) {
+    if (mUseTimeBracket && !isInsideTimeBracket(bracket)) {
       return true;
     }
 
-    if (mEnabledFilters.test(Filter::ITSROF) && !isInsideITSROF(bracket)) {
+    if (conf.filterITSROF && !isInsideITSROF(bracket)) {
       return true;
     }
 
@@ -218,36 +221,58 @@ void EveWorkflowHelper::selectTracks(const CalibObjectsConst* calib,
 
     // If the mode is disabled,
     // add every track to a symbolic "zero" primary vertex
-    if (!mPrimaryVertexMode) {
+    if (!conf.PVMode) {
       mPrimaryVertexTrackGIDs[0].push_back(gid);
     }
 
     return true;
   };
 
+  auto prop = o2::base::Propagator::Instance();
   this->mRecoCont->createTracksVariadic(creator);
-
-  if (mPrimaryVertexMode) {
+  if (conf.PVMode) {
+    bool checkTPCDCA = conf.TPCOnlyMaxDCARZ[0] > 0.f || conf.TPCOnlyMaxDCARZ[1] > 0.f;
     const auto trackIndex = mRecoCont->getPrimaryVertexMatchedTracks(); // Global ID's for associated tracks
     const auto vtxRefs = mRecoCont->getPrimaryVertexMatchedTrackRefs(); // references from vertex to these track IDs
     const auto totalPrimaryVertices = vtxRefs.size() - 1;               // The last entry is for unassigned tracks, ignore them
 
     for (std::size_t iv = 0; iv < totalPrimaryVertices; iv++) {
+      const auto& pv = mRecoCont->getPrimaryVertex(iv);
+      if (pv.getX() < conf.PVXYZMin[0] || pv.getX() > conf.PVXYZMax[0] ||
+          pv.getY() < conf.PVXYZMin[1] || pv.getY() > conf.PVXYZMax[1] ||
+          pv.getZ() < conf.PVXYZMin[2] || pv.getZ() > conf.PVXYZMax[2]) {
+        continue; // skip vertex
+      }
+
+      o2::math_utils::Point3D<float> pvXYZ(pv.getX(), pv.getY(), pv.getZ());
       const auto& vtref = vtxRefs[iv];
       const gsl::span tracksForVertex(trackIndex.data() + vtref.getFirstEntry(), vtref.getEntries());
       for (const auto& tvid : tracksForVertex) {
-
         if (!mRecoCont->isTrackSourceLoaded(tvid.getSource())) {
           continue;
         }
-
-        // TODO: fix TPC tracks?
-
-        GID gid = tvid;
+        GID gid(tvid);
         fixMFTMCHMIDLabel(gid);
-
         // If a track was not rejected, associate it with its primary vertex
         if (mGIDTrackTime.find(gid) != mGIDTrackTime.end()) {
+          // add optional cut on TPC-only tracks DCAz (corrected for the vertex time)
+          if (gid.getSource() == o2::dataformats::GlobalTrackID::TPC && checkTPCDCA) {
+            const auto& tpcTr = mRecoCont->getTPCTrack(gid);
+            o2::track::TrackPar trc{tpcTr};
+            if (!tpcTr.hasBothSidesClusters()) { // need to correct track Z with this vertex time
+              float dz = (tpcTr.getTime0() * mTPCTimeBins2MUS - (pv.getTimeStamp().getTimeStamp() + mTPCVDrift->getTimeOffset())) * mTPCVDrift->getVDrift();
+              if (tpcTr.hasCSideClustersOnly()) {
+                dz = -dz;
+              }
+              trc.setZ(trc.getZ() + dz);
+            }
+            std::array<float, 2> dca;
+            if (!prop->propagateToDCA(pvXYZ, trc, prop->getNominalBz(), 10., o2::base::PropagatorF::MatCorrType::USEMatCorrNONE, &dca) ||
+                (conf.TPCOnlyMaxDCARZ[1] > 0. && std::abs(dca[1]) > conf.TPCOnlyMaxDCARZ[1]) ||
+                (conf.TPCOnlyMaxDCARZ[0] > 0. && std::abs(dca[0]) > conf.TPCOnlyMaxDCARZ[0])) {
+              continue;
+            }
+          }
           mPrimaryVertexTrackGIDs[iv].push_back(gid);
         }
       }
@@ -260,22 +285,22 @@ void EveWorkflowHelper::selectTowers()
   const auto& allTriggersPHOS = mRecoCont->getPHOSTriggers();
   const auto& allTriggersEMCAL = mRecoCont->getEMCALTriggers();
   const auto& allTriggersHMP = mRecoCont->getHMPClusterTriggers();
-
+  const auto& conf = EveConfParam::Instance();
   auto filterTrigger = [&](const auto& trig) {
     const auto time = bcDiffToTFTimeMUS(trig.getBCData());
 
-    if (mEnabledFilters.test(Filter::TimeBracket) && !isInsideTimeBracket(time)) {
+    if (mUseTimeBracket && !isInsideTimeBracket(time)) {
       return false;
     }
 
-    if (mEnabledFilters.test(Filter::ITSROF) && !isInsideITSROF(time)) {
+    if (EveConfParam::Instance().filterITSROF && !isInsideITSROF(time)) {
       return false;
     }
 
     return true;
   };
 
-  if (mPrimaryVertexMode) {
+  if (conf.PVMode) {
     const auto trackIndex = mRecoCont->getPrimaryVertexMatchedTracks(); // Global ID's for associated tracks
     const auto vtxRefs = mRecoCont->getPrimaryVertexMatchedTrackRefs(); // references from vertex to these track IDs
     const auto totalPrimaryVertices = vtxRefs.size() - 1;               // The last entry is for unassigned tracks, ignore them
@@ -345,7 +370,7 @@ void EveWorkflowHelper::selectTowers()
 
 void EveWorkflowHelper::setITSROFs()
 {
-  if (mEnabledFilters.test(Filter::ITSROF)) {
+  if (EveConfParam::Instance().filterITSROF) {
     const auto& irFrames = mRecoCont->getIRFramesITS();
 
     for (const auto& irFrame : irFrames) {
@@ -356,6 +381,7 @@ void EveWorkflowHelper::setITSROFs()
 
 void EveWorkflowHelper::draw(std::size_t primaryVertexIdx, bool sortTracks)
 {
+  const auto& conf = EveConfParam::Instance();
   if (mPrimaryVertexTrackGIDs.find(primaryVertexIdx) == mPrimaryVertexTrackGIDs.end()) {
     LOGF(info, "Primary vertex %d has no tracks", primaryVertexIdx);
   } else {
@@ -364,6 +390,7 @@ void EveWorkflowHelper::draw(std::size_t primaryVertexIdx, bool sortTracks)
     };
 
     auto& tracks = mPrimaryVertexTrackGIDs.at(primaryVertexIdx);
+    float pvTime = conf.PVMode ? mRecoCont->getPrimaryVertex(primaryVertexIdx).getTimeStamp().getTimeStamp() : -1e9;
 
     if (sortTracks) {
       std::sort(tracks.begin(), tracks.end(),
@@ -374,8 +401,8 @@ void EveWorkflowHelper::draw(std::size_t primaryVertexIdx, bool sortTracks)
 
     auto trackCount = tracks.size();
 
-    if (mEnabledFilters.test(Filter::TotalNTracks) && trackCount >= mMaxNTracks) {
-      trackCount = mMaxNTracks;
+    if (conf.maxTracks > 0 && trackCount >= conf.maxTracks) {
+      trackCount = conf.maxTracks;
     }
 
     for (std::size_t it = 0; it < trackCount; it++) {
@@ -387,9 +414,20 @@ void EveWorkflowHelper::draw(std::size_t primaryVertexIdx, bool sortTracks)
         case GID::ITS:
           drawITS(gid, tim);
           break;
-        case GID::TPC:
-          drawTPC(gid, tim);
-          break;
+        case GID::TPC: {
+          float dz = 0.f;
+          if (conf.PVMode) { // for TPC the nominal time (center of the bracket) is stored but in the PVMode we correct it by the PV time
+            tim = pvTime;
+            const auto& tpcTr = mRecoCont->getTPCTrack(gid);
+            if (!tpcTr.hasBothSidesClusters()) { // need to correct track Z with this vertex time
+              float dz = (tpcTr.getTime0() * mTPCTimeBins2MUS - (pvTime + mTPCVDrift->getTimeOffset())) * mTPCVDrift->getVDrift();
+              if (tpcTr.hasCSideClustersOnly()) {
+                dz = -dz;
+              }
+            }
+          }
+          drawTPC(gid, tim, dz);
+        } break;
         case GID::MFT:
           drawMFT(gid, tim);
           break;
@@ -628,6 +666,7 @@ void EveWorkflowHelper::drawEMC(GID gid)
   const auto& trig = mRecoCont->getEMCALTriggers()[gid.getIndex()];
   const auto time = bcDiffToTFTimeMUS(trig.getBCData());
   const gsl::span cellsForTrigger(cells.data() + trig.getFirstEntry(), trig.getNumberOfObjects());
+  const auto& conf = EveConfParam::Instance();
 
   for (const auto& cell : cellsForTrigger) {
     if (!(cell.getType() == o2::emcal::ChannelType_t::HIGH_GAIN || cell.getType() == o2::emcal::ChannelType_t::LOW_GAIN)) {
@@ -646,11 +685,11 @@ void EveWorkflowHelper::drawEMC(GID gid)
       cellTime = calibCell->getTimeStamp();
       cellEnergy = calibCell->getEnergy();
     }
-    if (std::abs(cellTime) > mEMCALMaxCellTime) {
+    if (std::abs(cellTime) > conf.EMCCellTimeMax) {
       // cell rejected by time cut (pileup or noise)
       continue;
     }
-    if (cellEnergy < mEMCALMinCellEnergy) {
+    if (cellEnergy < conf.EMCCellEnergyMin) {
       // cell rejected by energy cut (rejection of soft cells)
       continue;
     }
@@ -681,7 +720,7 @@ void EveWorkflowHelper::drawITSTPC(GID gid, float trackTime, GID::Source source)
   const auto& track = mRecoCont->getTPCITSTrack(gid);
   addTrackToEvent(track, gid, trackTime, 0., source);
   drawITSClusters(track.getRefITS()); // trackTime
-  drawTPCClusters(track.getRefTPC()); // trackTime * mMUS2TPCTimeBins
+  drawTPCClusters(track.getRefTPC(), trackTime * mMUS2TPCTimeBins);
 }
 
 void EveWorkflowHelper::drawITSTPCTOF(GID gid, float trackTime, GID::Source source)
@@ -689,7 +728,7 @@ void EveWorkflowHelper::drawITSTPCTOF(GID gid, float trackTime, GID::Source sour
   const auto& track = mRecoCont->getITSTPCTOFTrack(gid);
   addTrackToEvent(track, gid, trackTime, 0., source);
   drawITSClusters(track.getRefITS()); // trackTime
-  drawTPCClusters(track.getRefTPC()); // trackTime * mMUS2TPCTimeBins
+  drawTPCClusters(track.getRefTPC(), trackTime * mMUS2TPCTimeBins);
   drawTOFClusters(gid);
 }
 
@@ -698,7 +737,7 @@ void EveWorkflowHelper::drawTPCTRD(GID gid, float trackTime, GID::Source source)
   // LOG(info) << "EveWorkflowHelper::drawTPCTRD " << gid;
   const auto& tpcTrdTrack = mRecoCont->getTPCTRDTrack<o2::trd::TrackTRD>(gid);
   addTrackToEvent(tpcTrdTrack, gid, trackTime, 0., source);
-  drawTPCClusters(tpcTrdTrack.getRefGlobalTrackId()); // trackTime * mMUS2TPCTimeBins
+  drawTPCClusters(tpcTrdTrack.getRefGlobalTrackId(), trackTime * mMUS2TPCTimeBins);
   drawTRDClusters(tpcTrdTrack);                       // tracktime
 }
 
@@ -734,7 +773,7 @@ void EveWorkflowHelper::drawTPCTOF(GID gid, float trackTime)
   const auto& trTPCTOF = mRecoCont->getTPCTOFTrack(gid);
   const auto& match = mRecoCont->getTPCTOFMatch(gid.getIndex());
   addTrackToEvent(trTPCTOF, gid, trackTime, 0);
-  drawTPCClusters(match.getTrackRef()); // trackTime * mMUS2TPCTimeBins
+  drawTPCClusters(match.getTrackRef(), trackTime * mMUS2TPCTimeBins);
   drawTOFClusters(gid);                 // trackTime
 }
 
@@ -937,19 +976,21 @@ void EveWorkflowHelper::drawITSClusters(GID gid) // float trackTime
 }
 
 // TPC cluseters for given TPC track (gid)
-void EveWorkflowHelper::drawTPCClusters(GID gid) // , float trackTimeTB
+void EveWorkflowHelper::drawTPCClusters(GID gid, float trackTimeTB)
 {
   const auto& trc = mRecoCont->getTPCTrack(gid);
   auto mTPCTracksClusIdx = mRecoCont->getTPCTracksClusterRefs();
   auto mTPCClusterIdxStruct = &mRecoCont->getTPCClusters();
-
+  if (trackTimeTB < -1e9) {
+    trackTimeTB = trc.getTime0();
+  }
   // store the TPC cluster positions
   for (int iCl = trc.getNClusterReferences(); iCl--;) {
     uint8_t sector, row;
     const auto& clTPC = trc.getCluster(mTPCTracksClusIdx, iCl, *mTPCClusterIdxStruct, sector, row);
 
     std::array<float, 3> xyz;
-    this->mTPCFastTransform->TransformIdeal(sector, row, clTPC.getPad(), clTPC.getTime(), xyz[0], xyz[1], xyz[2], trc.getTime0()); // in sector coordinate
+    this->mTPCFastTransform->TransformIdeal(sector, row, clTPC.getPad(), clTPC.getTime(), xyz[0], xyz[1], xyz[2], trackTimeTB);    // in sector coordinate
     o2::math_utils::rotateZ(xyz, o2::math_utils::sector2Angle(sector % o2::tpc::SECTORSPERSIDE));                                  // lab coordinate (global)
     mEvent.addCluster(xyz.data());                                                                                                 // trackTimeTB / mMUS2TPCTimeBins
   }
@@ -968,23 +1009,23 @@ void EveWorkflowHelper::drawMFTClusters(GID gid) // float trackTime
   }
 }
 
-void EveWorkflowHelper::drawTPC(GID gid, float trackTime)
+void EveWorkflowHelper::drawTPC(GID gid, float trackTime, float dz)
 {
   const auto& tr = mRecoCont->getTPCTrack(gid);
 
-  if (mEnabledFilters.test(Filter::EtaBracket) && mEtaBracket.isOutside(tr.getEta()) == Bracket::Relation::Inside) {
+  if (mUseEtaBracketTPC && mEtaBracketTPC.isOutside(tr.getEta()) == Bracket::Relation::Inside) {
     return;
   }
 
-  addTrackToEvent(tr, gid, trackTime, 4.f, GID::TPC);
-
-  drawTPCClusters(gid); // trackTime
+  addTrackToEvent(tr, gid, trackTime, dz, GID::TPC);
+  float clTime0 = EveConfParam::Instance().PVMode ? trackTime * mMUS2TPCTimeBins : -2e9; // in PVMode use supplied real time converted to TB, otherwise pass dummy time to use tpcTrack.getTime0
+  drawTPCClusters(gid, clTime0);                                                         // trackTime
 }
 
 void EveWorkflowHelper::drawITS(GID gid, float trackTime)
 {
   const auto& tr = mRecoCont->getITSTrack(gid);
-  addTrackToEvent(tr, gid, trackTime, 1.f, GID::ITS);
+  addTrackToEvent(tr, gid, trackTime, 0.f, GID::ITS);
 
   drawITSClusters(gid); // tracktime
 }
@@ -1095,8 +1136,17 @@ void EveWorkflowHelper::drawTRDClusters(const o2::trd::TrackTRD& tpcTrdTrack) //
   }
 }
 
-EveWorkflowHelper::EveWorkflowHelper(const FilterSet& enabledFilters, std::size_t maxNTracks, const Bracket& timeBracket, const Bracket& etaBracket, bool primaryVertexMode) : mEnabledFilters(enabledFilters), mMaxNTracks(maxNTracks), mTimeBracket(timeBracket), mEtaBracket(etaBracket), mPrimaryVertexMode(primaryVertexMode)
+EveWorkflowHelper::EveWorkflowHelper()
 {
+  const auto& conf = EveConfParam::Instance();
+  if (conf.timeMinMax[0] >= 0.f && conf.timeMinMax[0] < conf.timeMinMax[1]) {
+    mUseTimeBracket = true;
+    mTimeBracket = {conf.timeMinMax[0], conf.timeMinMax[1]};
+  }
+  if (conf.TPCEtaAbsMax > 0.f) {
+    mUseEtaBracketTPC = true;
+    mEtaBracketTPC = {-conf.TPCEtaAbsMax, conf.TPCEtaAbsMax};
+  }
   o2::mch::TrackExtrap::setField();
   this->mMFTGeom = o2::mft::GeometryTGeo::Instance();
   this->mMFTGeom->fillMatrixCache(o2::math_utils::bit2Mask(o2::math_utils::TransformType::T2L, o2::math_utils::TransformType::L2G));
@@ -1106,6 +1156,7 @@ EveWorkflowHelper::EveWorkflowHelper(const FilterSet& enabledFilters, std::size_
   this->mPHOSGeom = o2::phos::Geometry::GetInstance("");
   this->mTPCFastTransform = (o2::tpc::TPCFastTransformHelperO2::instance()->create(0));
   const auto& elParams = o2::tpc::ParameterElectronics::Instance();
+  mTPCTimeBins2MUS = elParams.ZbinWidth;
   mMUS2TPCTimeBins = 1. / elParams.ZbinWidth;
   mTPCBin2MUS = elParams.ZbinWidth;
   const auto grp = o2::base::GRPGeomHelper::instance().getGRPECS();
@@ -1119,6 +1170,14 @@ EveWorkflowHelper::EveWorkflowHelper(const FilterSet& enabledFilters, std::size_
 
   for (int i = 0; i < GID::Source::NSources; i++) {
     mTotalDataTypes[i] = 0;
+  }
+}
+
+void EveWorkflowHelper::setTPCVDrift(const o2::tpc::VDriftCorrFact* v)
+{
+  mTPCVDrift = v;
+  if (v) {
+    o2::tpc::TPCFastTransformHelperO2::instance()->updateCalibration(*mTPCFastTransform.get(), 0, v->corrFact, v->refVDrift, v->getTimeOffset());
   }
 }
 

--- a/EventVisualisation/Workflow/src/O2DPLDisplay.cxx
+++ b/EventVisualisation/Workflow/src/O2DPLDisplay.cxx
@@ -17,6 +17,7 @@
 #include "Framework/CompletionPolicyHelpers.h"
 #include "EveWorkflow/O2DPLDisplay.h"
 #include "EveWorkflow/EveWorkflowHelper.h"
+#include "EventVisualisationBase/EveConfParam.h"
 #include "EventVisualisationBase/ConfigurationManager.h"
 #include "DetectorsBase/Propagator.h"
 #include "DataFormatsGlobalTracking/RecoContainer.h"
@@ -63,37 +64,15 @@ void customize(std::vector<ConfigParamSpec>& workflowOptions)
     {"use-root-format", VariantType::Bool, false, {"instead of eve format (default) use root format"}},
     {"eve-hostname", VariantType::String, "", {"name of the host allowed to produce files (empty means no limit)"}},
     {"eve-dds-collection-index", VariantType::Int, -1, {"number of dpl collection allowed to produce files (-1 means no limit)"}},
-    {"number-of_files", VariantType::Int, 150, {"maximum number of json files in folder"}},
-    {"number-of_tracks", VariantType::Int, -1, {"maximum number of track stored in json file (-1 means no limit)"}},
-    {"number-of_bytes", VariantType::Int, 3000000, {"number of bytes stored in time interval which stops producing new data file (-1 means no limit)"}},
     {"time-interval", VariantType::Int, 5000, {"time interval in milliseconds between stored files"}},
     {"disable-mc", VariantType::Bool, false, {"disable visualization of MC data"}},
     {"disable-write", VariantType::Bool, false, {"disable writing output files"}},
     {"display-clusters", VariantType::String, "ITS,TPC,TRD,TOF", {"comma-separated list of clusters to display"}},
     {"display-tracks", VariantType::String, "TPC,ITS,ITS-TPC,TPC-TRD,ITS-TPC-TRD,TPC-TOF,ITS-TPC-TOF", {"comma-separated list of tracks to display"}},
     {"disable-root-input", VariantType::Bool, false, {"disable root-files input reader"}},
-    {"configKeyValues", VariantType::String, "", {"semicolon separated key=value strings ..."}},
+    {"configKeyValues", VariantType::String, "", {"semicolon separated key=value strings, e.g. EveConfParam content..."}},
     {"skipOnEmptyInput", VariantType::Bool, false, {"don't run the ED when no input is provided"}},
-    {"min-its-tracks", VariantType::Int, -1, {"don't create file if less than the specified number of ITS tracks is present"}},
-    {"min-tracks", VariantType::Int, 1, {"don't create file if less than the specified number of all tracks is present"}},
-    {"filter-its-rof", VariantType::Bool, false, {"don't display tracks outside ITS readout frame"}},
-    {"filter-time-min", VariantType::Float, -1.f, {"display tracks only in [min, max] microseconds time range in each time frame, requires --filter-time-max to be specified as well"}},
-    {"filter-time-max", VariantType::Float, -1.f, {"display tracks only in [min, max] microseconds time range in each time frame, requires --filter-time-min to be specified as well"}},
-    {"remove-tpc-abs-eta", VariantType::Float, 0.f, {"remove TPC tracks in [-eta, +eta] range"}},
-    {"track-sorting", VariantType::Bool, true, {"sort track by track time before applying filters"}},
-    {"only-nth-event", VariantType::Int, 0, {"process only every nth event"}},
-    {"primary-vertex-mode", VariantType::Bool, false, {"produce jsons with individual primary vertices, not total time frame data"}},
-    {"max-primary-vertices", VariantType::Int, 5, {"maximum number of primary vertices to draw per time frame"}},
-    {"primary-vertex-triggers", VariantType::Bool, false, {"instead of drawing vertices with tracks (and maybe calorimeter triggers), draw vertices with calorimeter triggers (and maybe tracks)"}},
-    {"no-calibrate-emcal", VariantType::Bool, false, {"Do not apply on-the-fly EMCAL calibration"}},
-    {"emcal-max-celltime", VariantType::Float, 100.f, {"Max. EMCAL cell time (in ns)"}},
-    {"emcal-min-cellenergy", VariantType::Float, 0.3f, {"Min. EMCAL cell energy (in GeV)"}},
-    {"primary-vertex-min-z", VariantType::Float, -o2::constants::math::VeryBig, {"minimum z position for primary vertex"}},
-    {"primary-vertex-max-z", VariantType::Float, o2::constants::math::VeryBig, {"maximum z position for primary vertex"}},
-    {"primary-vertex-min-x", VariantType::Float, -o2::constants::math::VeryBig, {"minimum x position for primary vertex"}},
-    {"primary-vertex-max-x", VariantType::Float, o2::constants::math::VeryBig, {"maximum x position for primary vertex"}},
-    {"primary-vertex-min-y", VariantType::Float, -o2::constants::math::VeryBig, {"minimum y position for primary vertex"}},
-    {"primary-vertex-max-y", VariantType::Float, o2::constants::math::VeryBig, {"maximum y position for primary vertex"}}};
+  };
 
   o2::raw::HBFUtilsInitializer::addConfigOption(options);
   std::swap(workflowOptions, options);
@@ -112,10 +91,12 @@ void O2DPLDisplaySpec::init(InitContext& ic)
 
 void O2DPLDisplaySpec::run(ProcessingContext& pc)
 {
-  if (!this->mEveHostNameMatch) {
+  const auto& conf = EveConfParam::Instance();
+
+  if (!mEveHostNameMatch) {
     return;
   }
-  if (this->mOnlyNthEvent && this->mEventCounter++ % this->mOnlyNthEvent != 0) {
+  if (conf.onlyNthEvent > 1 && mEventCounter++ % conf.onlyNthEvent) {
     return;
   }
   LOGF(info, "------------------------    O2DPLDisplay::run version ", o2_eve_version, "    ------------------------------------");
@@ -142,19 +123,12 @@ void O2DPLDisplaySpec::run(ProcessingContext& pc)
     }
   }
 
-  EveWorkflowHelper::FilterSet enabledFilters;
-
-  enabledFilters.set(EveWorkflowHelper::Filter::ITSROF, this->mFilterITSROF);
-  enabledFilters.set(EveWorkflowHelper::Filter::TimeBracket, this->mFilterTime);
-  enabledFilters.set(EveWorkflowHelper::Filter::EtaBracket, this->mRemoveTPCEta);
-  enabledFilters.set(EveWorkflowHelper::Filter::TotalNTracks, this->mNumberOfTracks != -1);
-  EveWorkflowHelper helper(enabledFilters, this->mNumberOfTracks, this->mTimeBracket, this->mEtaBracket, this->mPrimaryVertexMode);
+  EveWorkflowHelper helper;
+  helper.setTPCVDrift(&mTPCVDriftHelper.getVDriftObject());
   helper.setRecoContainer(&recoCont);
   if (mEMCALCalibrator) {
     helper.setEMCALCellRecalibrator(mEMCALCalibrator.get());
   }
-  helper.setMaxEMCALCellTime(mEMCALMaxCellTime);
-  helper.setMinEMCALCellEnergy(mEMCALMinCellEnergy);
 
   helper.setITSROFs();
   helper.selectTracks(&(mData.mConfig.configCalib), mClMask, mTrkMask, mTrkMask);
@@ -167,43 +141,35 @@ void O2DPLDisplaySpec::run(ProcessingContext& pc)
   std::size_t filesSaved = 0;
   const std::vector<std::string> dirs = o2::event_visualisation::DirectoryLoader::allFolders(this->mJsonPath);
   const std::string marker = "_";
-  const std::vector<std::string> exts = {
-    ".json", ".root", ".eve"};
+  const std::vector<std::string> exts = {".json", ".root", ".eve"};
   auto processData = [&](const auto& dataMap) {
     for (const auto& keyVal : dataMap) {
-      if (filesSaved >= mMaxPrimaryVertices) {
+      if (conf.maxPVs > 0 && filesSaved >= conf.maxPVs) {
         break;
       }
-      if (this->mNumberOfBytes != -1) {
+      if (conf.maxBytes > 0) {
         auto periodStart =
           duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() - this->mTimeInterval.count();
-        if (!DirectoryLoader::canCreateNextFile(
-              dirs, marker, exts, periodStart, this->mNumberOfBytes)) {
-          LOGF(info, "Already too much data (> %d) to transfer in this period - event will not be not saved ...", this->mNumberOfBytes);
+        if (!DirectoryLoader::canCreateNextFile(dirs, marker, exts, periodStart, conf.maxBytes)) {
+          LOGF(info, "Already too much data (> %d) to transfer in this period - event will not be not saved ...", conf.maxBytes);
           break;
         }
       }
       const auto pv = keyVal.first;
       bool save = false;
-      if (mPrimaryVertexMode) {
-        auto primaryVertex = recoCont.getPrimaryVertices()[pv];
-        auto primaryVertex_X = primaryVertex.getX();
-        auto primaryVertex_Y = primaryVertex.getY();
-        auto primaryVertex_Z = primaryVertex.getZ();
-        if ((primaryVertex_X >= mPrimaryVertexMinX) & (primaryVertex_X <= mPrimaryVertexMaxX) & (primaryVertex_Y >= mPrimaryVertexMinY) & (primaryVertex_Y <= mPrimaryVertexMaxY) & (primaryVertex_Z >= mPrimaryVertexMinZ) & (primaryVertex_Z <= mPrimaryVertexMaxZ)) {
-          helper.draw(pv, mTrackSorting);
-          save = true;
-        }
+      if (conf.PVMode) {
+        helper.draw(pv, conf.trackSorting);
+        save = true;
       } else {
-        helper.draw(pv, mTrackSorting);
+        helper.draw(pv, conf.trackSorting);
         save = true;
       }
 
-      if (this->mMinITSTracks != -1 && helper.mEvent.getDetectorTrackCount(detectors::DetID::ITS) < this->mMinITSTracks) {
+      if (conf.minITSTracks > -1 && helper.mEvent.getDetectorTrackCount(detectors::DetID::ITS) < conf.minITSTracks) {
         save = false;
       }
 
-      if (this->mMinTracks != -1 && helper.mEvent.getTrackCount() < this->mMinTracks) {
+      if (conf.minTracks > -1 && helper.mEvent.getTrackCount() < conf.minTracks) {
         save = false;
       }
 
@@ -220,7 +186,7 @@ void O2DPLDisplaySpec::run(ProcessingContext& pc)
         helper.mEvent.setRunType(this->mRunType);
         helper.mEvent.setPrimaryVertex(pv);
         helper.mEvent.setCreationTime(tinfo.creation);
-        helper.save(this->mJsonPath, this->mExt, this->mNumberOfFiles);
+        helper.save(this->mJsonPath, this->mExt, conf.maxFiles);
         filesSaved++;
         currentTime = std::chrono::high_resolution_clock::now(); // time AFTER save
         this->mTimeStamp = currentTime;                          // next run AFTER period counted from last save
@@ -229,7 +195,7 @@ void O2DPLDisplaySpec::run(ProcessingContext& pc)
       helper.clear();
     }
   };
-  if (mPrimaryVertexTriggers) {
+  if (conf.PVTriggersMode) {
     processData(helper.mPrimaryVertexTriggerGIDs);
   } else {
     processData(helper.mPrimaryVertexTrackGIDs);
@@ -274,6 +240,10 @@ void O2DPLDisplaySpec::updateTimeDependentParams(ProcessingContext& pc)
 {
   o2::base::GRPGeomHelper::instance().checkUpdates(pc);
   static bool initOnceDone = false;
+  mTPCVDriftHelper.extractCCDBInputs(pc);
+  if (mTPCVDriftHelper.isUpdated()) {
+    mTPCVDriftHelper.acknowledgeUpdate();
+  }
   if (!initOnceDone) { // this params need to be queried only once
     initOnceDone = true;
     auto grpECS = o2::base::GRPGeomHelper::instance().getGRPECS(); // RS
@@ -302,6 +272,10 @@ void O2DPLDisplaySpec::finaliseCCDB(ConcreteDataMatcher& matcher, void* obj)
     mData.setMFTDict((const o2::itsmft::TopologyDictionary*)obj);
     return;
   }
+  if (mTPCVDriftHelper.accountCCDBInputs(matcher, obj)) {
+
+    return;
+  }
 }
 
 WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
@@ -322,6 +296,8 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
   }
   auto eveHostName = cfgc.options().get<std::string>("eve-hostname");
   o2::conf::ConfigurableParam::updateFromString(cfgc.options().get<std::string>("configKeyValues"));
+  const auto& conf = EveConfParam::Instance();
+
   bool useMC = !cfgc.options().get<bool>("disable-mc");
   bool disableWrite = cfgc.options().get<bool>("disable-write");
 
@@ -342,9 +318,6 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
   }
 
   std::chrono::milliseconds timeInterval(cfgc.options().get<int>("time-interval"));
-  int numberOfFiles = cfgc.options().get<int>("number-of_files");
-  int numberOfTracks = cfgc.options().get<int>("number-of_tracks");
-  int numberOfBytes = cfgc.options().get<int>("number-of_bytes");
 
   GlobalTrackID::mask_t srcTrk = GlobalTrackID::getSourcesMask(cfgc.options().get<std::string>("display-tracks"));
   GlobalTrackID::mask_t srcCl = GlobalTrackID::getSourcesMask(cfgc.options().get<std::string>("display-clusters"));
@@ -386,55 +359,22 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
     return std::make_tuple(optEnabled, bracket);
   };
 
-  const auto [filterTime, timeBracket] = isRangeEnabled("filter-time-min", "filter-time-max");
-
-  const auto etaRange = cfgc.options().get<float>("remove-tpc-abs-eta");
-
-  bool removeTPCEta = false;
-  EveWorkflowHelper::Bracket etaBracket;
-
-  if (etaRange != 0.f) {
-    etaBracket = EveWorkflowHelper::Bracket{-etaRange, etaRange};
-    removeTPCEta = true;
-  }
-
   std::shared_ptr<DataRequest> dataRequest = std::make_shared<DataRequest>();
   dataRequest->requestTracks(srcTrk, useMC);
   dataRequest->requestClusters(srcCl, useMC);
 
-  auto filterITSROF = cfgc.options().get<bool>("filter-its-rof");
-
-  if (filterITSROF) {
+  if (conf.filterITSROF) {
     dataRequest->requestIRFramesITS();
     InputHelper::addInputSpecsIRFramesITS(cfgc, specs);
   }
 
-  auto primaryVertexMode = cfgc.options().get<bool>("primary-vertex-mode");
-  auto maxPrimaryVertices = cfgc.options().get<int>("max-primary-vertices");
-
   InputHelper::addInputSpecs(cfgc, specs, srcCl, srcTrk, srcTrk, useMC);
-  if (primaryVertexMode) {
+  if (conf.PVMode) {
     dataRequest->requestPrimaryVertices(useMC);
     InputHelper::addInputSpecsPVertex(cfgc, specs, useMC);
   }
+  o2::tpc::VDriftHelper::requestCCDBInputs(dataRequest->inputs);
 
-  auto minITSTracks = cfgc.options().get<int>("min-its-tracks");
-  auto minTracks = cfgc.options().get<int>("min-tracks");
-  auto onlyNthEvent = cfgc.options().get<int>("only-nth-event");
-  auto tracksSorting = cfgc.options().get<bool>("track-sorting");
-  auto primaryVertexTriggers = cfgc.options().get<bool>("primary-vertex-triggers");
-  auto primaryVertexMinZ = cfgc.options().get<float>("primary-vertex-min-z");
-  auto primaryVertexMaxZ = cfgc.options().get<float>("primary-vertex-max-z");
-  auto primaryVertexMinX = cfgc.options().get<float>("primary-vertex-min-x");
-  auto primaryVertexMaxX = cfgc.options().get<float>("primary-vertex-max-x");
-  auto primaryVertexMinY = cfgc.options().get<float>("primary-vertex-min-y");
-  auto primaryVertexMaxY = cfgc.options().get<float>("primary-vertex-max-y");
-  auto maxEMCALCellTime = cfgc.options().get<float>("emcal-max-celltime");
-  auto minEMCALCellEnergy = cfgc.options().get<float>("emcal-min-cellenergy");
-
-  if (numberOfTracks == -1) {
-    tracksSorting = false; // do not sort if all tracks are allowed
-  }
   auto ggRequest = std::make_shared<o2::base::GRPGeomRequest>(false,                             // orbitResetTime
                                                               true,                              // GRPECS=true
                                                               false,                             // GRPLHCIF
@@ -445,7 +385,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
                                                               true); // query only once all objects except mag.field
 
   std::shared_ptr<o2::emcal::CalibLoader> emcalCalibLoader;
-  if (!cfgc.options().get<bool>("no-calibrate-emcal")) {
+  if (conf.calibrateEMC) {
     emcalCalibLoader = std::make_shared<o2::emcal::CalibLoader>();
     emcalCalibLoader->enableTimeCalib(true);
     emcalCalibLoader->enableBadChannelMap(true);
@@ -457,7 +397,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
     "o2-eve-export",
     dataRequest->inputs,
     {},
-    AlgorithmSpec{adaptFromTask<O2DPLDisplaySpec>(disableWrite, useMC, srcTrk, srcCl, dataRequest, ggRequest, emcalCalibLoader, jsonFolder, ext, timeInterval, numberOfFiles, numberOfTracks, numberOfBytes, eveHostNameMatch, minITSTracks, minTracks, filterITSROF, filterTime, timeBracket, removeTPCEta, etaBracket, tracksSorting, onlyNthEvent, primaryVertexMode, maxPrimaryVertices, primaryVertexTriggers, primaryVertexMinZ, primaryVertexMaxZ, primaryVertexMinX, primaryVertexMaxX, primaryVertexMinY, primaryVertexMaxY, maxEMCALCellTime, minEMCALCellEnergy)}});
+    AlgorithmSpec{adaptFromTask<O2DPLDisplaySpec>(disableWrite, useMC, srcTrk, srcCl, dataRequest, ggRequest, emcalCalibLoader, jsonFolder, ext, timeInterval, eveHostNameMatch)}});
 
   // configure dpl timer to inject correct firstTForbit: start from the 1st orbit of TF containing 1st sampled orbit
   o2::raw::HBFUtilsInitializer hbfIni(cfgc, specs);

--- a/prodtests/full-system-test/dpl-workflow.sh
+++ b/prodtests/full-system-test/dpl-workflow.sh
@@ -81,7 +81,6 @@ TRD_CONFIG=
 TRD_CONFIG_KEY=
 TRD_FILTER_CONFIG=
 CPV_INPUT=raw
-EVE_CONFIG=" --jsons-folder $EDJSONS_DIR"
 MIDDEC_CONFIG=
 EMCRAW2C_CONFIG=
 PHS_CONFIG=
@@ -90,6 +89,8 @@ CTP_CONFIG=
 TPC_CORR_OPT=
 TPC_CORR_KEY=
 INTERACTION_TAG_CONFIG_KEY=
+EVE_OPT=" --jsons-folder $EDJSONS_DIR"
+: ${EVE_CONFIG:=}
 : ${STRTRACKING:=}
 : ${ITSEXTRAERR:=}
 : ${TRACKTUNETPCINNER:=}
@@ -202,7 +203,7 @@ has_detector_flp_processing CPV && CPV_INPUT=digits
 ! has_detector_flp_processing TOF && TOF_CONFIG+=" --local-cmp"
 
 if [[ $EPNSYNCMODE == 1 ]]; then
-  EVE_CONFIG+=" --eve-dds-collection-index 0"
+  EVE_OPT+=" --eve-dds-collection-index 0"
   MIDDEC_CONFIG+=" --feeId-config-file \"$MID_FEEID_MAP\""
   if [[ $EXTINPUT == 1 ]] && [[ $GPUTYPE != "CPU" ]] && [[ -z "$GPU_NUM_MEM_REG_CALLBACKS" ]]; then
     if [[ $NUMAGPUIDS == 1 ]]; then
@@ -225,17 +226,18 @@ if [[ $SYNCRAWMODE == 1 ]] || [[ $SYNCMODE == 0 && $CTFINPUT == 1 && $GPUTYPE !=
 fi
 
 if [[ $SYNCMODE == 1 && "0${ED_NO_ITS_ROF_FILTER:-}" != "01" && $BEAMTYPE == "PbPb" ]] && has_detector ITS; then
-  EVE_CONFIG+=" --filter-its-rof"
+  EVE_CONFIG+=";eveconf.filterITSROF=true;"
 fi
 
 if [[ ! -z ${EVE_NTH_EVENT:-} ]]; then
-  EVE_CONFIG+=" --only-nth-event=$EVE_NTH_EVENT"
+  EVE_CONFIG+=";eveconf.onlyNthEvent=$EVE_NTH_EVENT;"
 fi
 
+EVE_CONFIG+=";eveconf.maxTracks=5000;"
 if [[ $RUNTYPE == "SYNTHETIC" ]]; then
-  EVE_CONFIG+=" --number-of_files 20"
+  EVE_CONFIG+=";eveconf.maxFiles=20;"
 elif [[ $RUNTYPE == "PHYSICS" ]]; then
-  EVE_CONFIG+=" --primary-vertex-triggers"
+  EVE_CONFIG+=";eveconf.PVTriggersMode=true;"
 fi
 
 if [[ $GPUTYPE != "CPU" && $NUMAGPUIDS != 0 ]] && [[ -z ${ROCR_VISIBLE_DEVICES:-} || ${ROCR_VISIBLE_DEVICES:-} = "0,1,2,3,4,5,6,7" || ${ROCR_VISIBLE_DEVICES:-} = "0,1,2,3" || ${ROCR_VISIBLE_DEVICES:-} = "4,5,6,7" ]]; then
@@ -352,7 +354,7 @@ if has_processing_step MUON_SYNC_RECO; then
   [[ $RUNTYPE == "COSMICS" ]] && [[ -z ${CONFIG_EXTRA_PROCESS_o2_mft_reco_workflow:-} ]] && CONFIG_EXTRA_PROCESS_o2_mft_reco_workflow="MFTTracking.FullClusterScan=true"
 fi
 [[ $SYNCRAWMODE == 1 ]] && [[ -z ${CONFIG_EXTRA_PROCESS_o2_zdc_digits_reco:-} ]] && CONFIG_EXTRA_PROCESS_o2_zdc_digits_reco='RecoParamZDC.tdc_calib[9]=1;RecoParamZDC.tdc_calib[0]=1;RecoParamZDC.tdc_calib[8]=1;RecoParamZDC.tdc_calib[1]=1;RecoParamZDC.tdc_calib[3]=1;RecoParamZDC.tdc_calib[6]=1;RecoParamZDC.tdc_calib[5]=1;RecoParamZDC.tdc_calib[4]=1;RecoParamZDC.tdc_calib[2]=1;RecoParamZDC.tdc_calib[7]=1;RecoParamZDC.energy_calib[13]=1;RecoParamZDC.energy_calib[12]=1;RecoParamZDC.energy_calib[11]=1;RecoParamZDC.energy_calib[6]=1;RecoParamZDC.energy_calib[25]=1;RecoParamZDC.energy_calib[14]=1;RecoParamZDC.energy_calib[20]=1;RecoParamZDC.energy_calib[5]=1;RecoParamZDC.energy_calib[0]=1;RecoParamZDC.energy_calib[19]=1;RecoParamZDC.tower_calib[1]=1;RecoParamZDC.tower_calib[2]=1;RecoParamZDC.tower_calib[3]=1;RecoParamZDC.tower_calib[4]=1;RecoParamZDC.tower_calib[24]=1;RecoParamZDC.tower_calib[21]=1;RecoParamZDC.tower_calib[22]=1;RecoParamZDC.tower_calib[23]=1;RecoParamZDC.tower_calib[18]=1;RecoParamZDC.tower_calib[16]=1;RecoParamZDC.tower_calib[17]=1;RecoParamZDC.tower_calib[15]=1;RecoParamZDC.tower_calib[8]=1;RecoParamZDC.tower_calib[9]=1;RecoParamZDC.tower_calib[7]=1;RecoParamZDC.tower_calib[10]=1'
-[[ $RUNTYPE != "COSMICS" ]] && [[ $RUNTYPE != "TECHNICAL" ]] && has_detectors_reco ITS && has_detector_matching PRIMVTX && [[ ! -z "$VERTEXING_SOURCES" ]] && EVE_CONFIG+=" --primary-vertex-mode"
+[[ $RUNTYPE != "COSMICS" ]] && [[ $RUNTYPE != "TECHNICAL" ]] && has_detectors_reco ITS && has_detector_matching PRIMVTX && [[ ! -z "$VERTEXING_SOURCES" ]] && EVE_CONFIG+=";eveconf.PVMode=true;"
 [[ $SYNCRAWMODE == 1 ]] && [[ -z ${CONFIG_EXTRA_PROCESS_o2_trd_global_tracking:-} ]] && CONFIG_EXTRA_PROCESS_o2_trd_global_tracking='GPU_rec_trd.maxChi2=25;GPU_rec_trd.penaltyChi2=20;GPU_rec_trd.extraRoadY=4;GPU_rec_trd.extraRoadZ=10;GPU_rec_trd.applyDeflectionCut=0;GPU_rec_trd.trkltResRPhiIdeal=1'
 [[ $SYNCRAWMODE == 1 ]] && [[ -z ${ARGS_EXTRA_PROCESS_o2_phos_reco_workflow:-} ]] && ARGS_EXTRA_PROCESS_o2_phos_reco_workflow='--presamples 2 --fitmethod semigaus'
 [[ $SYNCRAWMODE == 1 ]] && [[ $BEAMTYPE == "PbPb" ]] && [[ -z ${CONFIG_EXTRA_PROCESS_o2_calibration_emcal_channel_calib_workflow:-} ]] && CONFIG_EXTRA_PROCESS_o2_calibration_emcal_channel_calib_workflow='EMCALCalibParams.selectedClassMasks=C0TVX-NONE-NOPF-EMC:c0tvxtsc-b-nopf-emc:C0TVXTCE-B-NOPF-EMC;EMCALCalibParams.fractionEvents_bc=0.3'
@@ -628,7 +630,7 @@ workflow_has_parameters CALIB CALIB_LOCAL_INTEGRATED_AGGREGATOR && { source ${CA
 # RS this is a temporary setting
 : ${ED_TRACKS:=$TRACK_SOURCES}
 : ${ED_CLUSTERS:=$TRACK_SOURCES}
-workflow_has_parameter EVENT_DISPLAY && [[ $NUMAID == 0 ]] && [[ ! -z "$ED_TRACKS" ]] && [[ ! -z "$ED_CLUSTERS" ]] && [[ $EPNSYNCMODE == 0 || ${EPN_NODE_MI100:-0} == 0 ]] && add_W o2-eve-export-workflow "--display-tracks $ED_TRACKS --display-clusters $ED_CLUSTERS --skipOnEmptyInput $DISABLE_ROOT_INPUT --number-of_tracks 50000 $EVE_CONFIG $DISABLE_MC" "$ITSMFT_STROBES"
+workflow_has_parameter EVENT_DISPLAY && [[ $NUMAID == 0 ]] && [[ ! -z "$ED_TRACKS" ]] && [[ ! -z "$ED_CLUSTERS" ]] && [[ $EPNSYNCMODE == 0 || ${EPN_NODE_MI100:-0} == 0 ]] && add_W o2-eve-export-workflow "--display-tracks $ED_TRACKS --display-clusters $ED_CLUSTERS --skipOnEmptyInput $DISABLE_ROOT_INPUT $EVE_OPT $DISABLE_MC" "$ITSMFT_STROBES;$EVE_CONFIG"
 
 workflow_has_parameter GPU_DISPLAY && [[ $NUMAID == 0 ]] && add_W o2-gpu-display "${ED_TRACKS+--display-tracks} $ED_TRACKS ${ED_CLUSTERS+--display-clusters} $ED_CLUSTERS"
 

